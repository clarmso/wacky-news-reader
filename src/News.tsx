import React from "react";
import { Grid, Card, CardContent, Typography, Link } from "@material-ui/core";
import { makeStyles } from "@material-ui/core/styles";
import newgif from "./img/new.gif";

export type NewsItemProps = {
  title: string;
  byline: string;
  abstract: string;
  updated_date: string;
  url: string;
  is90s?: boolean;
  uri?: string;
};

const useStyles = makeStyles({
  card: {
    width: 400,
    height: 340,
  },
});

const NewsItem: React.FC<NewsItemProps> = ({
  title,
  byline,
  updated_date,
  abstract,
  url,
  is90s,
}) => {
  const classes = useStyles();
  const formattedDate = new Intl.DateTimeFormat("default", {
    hour: "numeric",
    minute: "numeric",
    day: "numeric",
    month: "long",
    year: "numeric",
    weekday: "long",
  }).format(Date.parse(updated_date));
  return (
    <Grid item>
      <Card className={classes.card}>
        <CardContent>
          <Typography variant="h5" component="h2" gutterBottom={true}>
            {is90s && <img src={newgif} alt="New" />}
            <Link href={url} rel="noreferrer" target="_blank" color="inherit">
              {title}
            </Link>
          </Typography>
          <Typography variant="subtitle2" gutterBottom={true}>
            {byline}
          </Typography>
          <Typography variant="subtitle2" gutterBottom={true}>
            {formattedDate}
          </Typography>
          <Typography variant="body1">{abstract}</Typography>
        </CardContent>
      </Card>
    </Grid>
  );
};

type NewsProps = {
  data: NewsItemProps[];
  is90s: boolean;
};

const News: React.FC<NewsProps> = ({ data, is90s }) => {
  return (
    <Grid container spacing={2} justify="center">
      {data.map((d: NewsItemProps) => {
        return (
          <NewsItem
            title={d.title}
            byline={d.byline}
            updated_date={d.updated_date}
            abstract={d.abstract}
            url={d.url}
            is90s={is90s}
            key={d.uri}
          />
        );
      })}
    </Grid>
  );
};

News.defaultProps = {
  data: [],
  is90s: false,
};

export default News;
