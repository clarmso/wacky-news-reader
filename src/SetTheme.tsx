import { useState } from "react";
import { useDispatch, useSelector } from "react-redux";

import {
  Grid,
  Radio,
  RadioGroup,
  FormControl,
  FormLabel,
  FormControlLabel,
} from "@material-ui/core";

import { set90sMode, setLightMode, setDarkMode } from "./connect/actions";
import { WackyState } from "./connect/reducer";

export const themeChoices = {
  LIGHT: "light",
  DARK: "dark",
  NINETYS: "90s",
};

const SetTheme: React.FC = () => {
  const theme = useSelector((state: WackyState) => state.themeChoice);
  const [themeChoice, setThemeChoice] = useState(theme);

  const dispatch = useDispatch();

  const handleChangeTheme = (event: React.ChangeEvent<HTMLInputElement>) => {
    const selected = event.target.value;
    switch (selected) {
      case themeChoices.NINETYS:
        dispatch(set90sMode());
        setThemeChoice(themeChoices.NINETYS);
        break;
      case themeChoices.DARK:
        dispatch(setDarkMode());
        setThemeChoice(themeChoices.DARK);
        break;
      case themeChoices.LIGHT:
        dispatch(setLightMode());
        setThemeChoice(themeChoices.LIGHT);
        break;
      default:
        dispatch(setLightMode());
        setThemeChoice(themeChoices.LIGHT);
    }
  };

  return (
    <Grid item xs>
      <FormControl>
        <FormLabel component="legend">Mode</FormLabel>
        <RadioGroup
          name="modes"
          value={themeChoice}
          onChange={handleChangeTheme}
          row
        >
          <FormControlLabel
            value={themeChoices.LIGHT}
            control={<Radio color="primary" />}
            label="Light"
            data-cy="light-mode"
          />
          <FormControlLabel
            value={themeChoices.DARK}
            control={<Radio color="primary" />}
            label="Dark"
            data-cy="dark-mode"
          />
          <FormControlLabel
            value={themeChoices.NINETYS}
            control={<Radio color="primary" />}
            label="90s ðŸŽ‰"
            data-cy="90s-mode"
          />
        </RadioGroup>
      </FormControl>
    </Grid>
  );
};

export default SetTheme;
