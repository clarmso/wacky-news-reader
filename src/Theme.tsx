import backgroundImage90s from "./img/pattern90s.png";

export const lightTheme = {
  overrides: {
    MuiCssBaseline: {
      "@global": {
        body: {
          backgroundImage: "none",
        },
      },
    },
  },
};

export const darkTheme = {
  palette: {
    type: "dark",
  },
  overrides: {
    MuiCssBaseline: {
      "@global": {
        body: {
          backgroundImage: "none",
        },
      },
    },
  },
};

export const NinetysTheme = {
  palette: {
    background: {
      paper: "hotpink",
      default: "hotpink",
    },
  },
  typography: {
    fontFamily: "Comic Sans MS",
  },
  overrides: {
    MuiCssBaseline: {
      "@global": {
        body: {
          backgroundImage: `url(${backgroundImage90s})`,
        },
      },
    },
  },
};
