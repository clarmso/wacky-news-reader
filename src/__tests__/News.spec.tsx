import { Provider } from "react-redux";
import { createStore } from "redux";
import { render } from "@testing-library/react";
import News from "../News";
import wackyReducer, { initialState } from "../connect/reducer";
import { themeChoices } from "../SetTheme";
import { ninetysTheme, lightTheme } from "../Theme";

it("renders with initial state", () => {
  const store = createStore(wackyReducer, initialState);
  const { asFragment } = render(
    <Provider store={store}>
      <News />
    </Provider>
  );
  const renderedHtml = asFragment();
  expect(renderedHtml).toMatchSnapshot();
});

it("renders with 90s mode", () => {
  const store = createStore(wackyReducer, {
    ...initialState,
    is90s: true,
    theme: ninetysTheme,
    themeChoice: themeChoices.NINETYS,
    news: [
      {
        title: "15 New Books to Watch For in May",
        byline: "By Joumana Khatib",
        abstract:
          "Buzzy new novels from Stacey Abrams, Jean Hanff Korelitz and Andy Weir; Michael Lewis’s take on the pandemic; Chimamanda Ngozi Adichie’s ode to grief and more.",
        updated_date: "2021-04-30T22:45:16-04:00",
        is90s: true,
        url: "https://www.nytimes.com/2021/04/28/books/may-2021-new-books.html",
        uri: "nyt://article/582af77c-6524-5a5c-8549-1536a31e8b82",
      },
    ],
  });
  const { asFragment } = render(
    <Provider store={store}>
      <News />
    </Provider>
  );
  const renderedHtml = asFragment();
  expect(renderedHtml).toMatchSnapshot();
});

it("renders with one news item", () => {
  const store = createStore(wackyReducer, {
    ...initialState,
    is90s: false,
    theme: lightTheme,
    themeChoice: themeChoices.LIGHT,
    news: [
      {
        title: "15 New Books to Watch For in May",
        byline: "By Joumana Khatib",
        abstract:
          "Buzzy new novels from Stacey Abrams, Jean Hanff Korelitz and Andy Weir; Michael Lewis’s take on the pandemic; Chimamanda Ngozi Adichie’s ode to grief and more.",
        updated_date: "2021-04-30T22:45:16-04:00",
        is90s: false,
        url: "https://www.nytimes.com/2021/04/28/books/may-2021-new-books.html",
        uri: "nyt://article/582af77c-6524-5a5c-8549-1536a31e8b82",
      },
    ],
  });
  const { asFragment } = render(
    <Provider store={store}>
      <News />
    </Provider>
  );
  const renderedHtml = asFragment();
  expect(renderedHtml).toMatchSnapshot();
});
