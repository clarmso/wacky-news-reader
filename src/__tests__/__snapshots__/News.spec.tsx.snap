// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with 90s mode 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-justify-xs-center"
  >
    <div
      class="MuiGrid-root MuiGrid-item"
    >
      <div
        class="MuiPaper-root MuiCard-root makeStyles-card-1 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="MuiCardContent-root"
        >
          <h2
            class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
          >
            <img
              alt="New"
              src="new.gif"
            />
            <a
              class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit"
              href="https://www.nytimes.com/2021/04/28/books/may-2021-new-books.html"
              rel="noreferrer"
              target="_blank"
            >
              15 New Books to Watch For in May
            </a>
          </h2>
          <h6
            class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-gutterBottom"
          >
            By Joumana Khatib
          </h6>
          <h6
            class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-gutterBottom"
          >
            Friday, April 30, 2021, 10:45 PM
          </h6>
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            Buzzy new novels from Stacey Abrams, Jean Hanff Korelitz and Andy Weir; Michael Lewis’s take on the pandemic; Chimamanda Ngozi Adichie’s ode to grief and more.
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders with initial state 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-justify-xs-center"
  />
</DocumentFragment>
`;

exports[`renders with one news item 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-justify-xs-center"
  >
    <div
      class="MuiGrid-root MuiGrid-item"
    >
      <div
        class="MuiPaper-root MuiCard-root makeStyles-card-2 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="MuiCardContent-root"
        >
          <h2
            class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
          >
            <a
              class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit"
              href="https://www.nytimes.com/2021/04/28/books/may-2021-new-books.html"
              rel="noreferrer"
              target="_blank"
            >
              15 New Books to Watch For in May
            </a>
          </h2>
          <h6
            class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-gutterBottom"
          >
            By Joumana Khatib
          </h6>
          <h6
            class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-gutterBottom"
          >
            Friday, April 30, 2021, 10:45 PM
          </h6>
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            Buzzy new novels from Stacey Abrams, Jean Hanff Korelitz and Andy Weir; Michael Lewis’s take on the pandemic; Chimamanda Ngozi Adichie’s ode to grief and more.
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
